frontend/src/Components/ScaleSelector/ScaleSelector.test.tsx

```diff
@@ -0,0 +1,275 @@
+
+    // when
+    const changeScaleButton = getByRole('button', { name: 'Change Scale' });
+    fireEvent.click(changeScaleButton);
+    const dropdown = getDropdown(container);
+    fireEvent.mouseMove(getByText('Fibonacci'));
+    fireEvent.mouseLeave(getByText('Fibonacci'));
+    fireEvent.keyDown(dropdown, { key: 'Enter' });
+
+    // then
+    expect(setScale).not.toHaveBeenCalled();
+    assertDropdownIsClosed(container);
+    expect(changeScaleButton).toHaveFocus();
+  });
+
+  it('closes the popup without selection when escape is pressed', () => {
+    // given
+    const setScale = jest.fn();
+    const { container, getByRole } = render({ setScale });
+
+    // when
+    const changeScaleButton = getByRole('button', { name: 'Change Scale' });
+    fireEvent.click(changeScaleButton);
+    const dropdown = getDropdown(container);
+    fireEvent.keyDown(dropdown, { key: 'Escape' });
+
+    // then
+    expect(setScale).not.toHaveBeenCalled();
+    assertDropdownIsClosed(container);
+    expect(changeScaleButton).toHaveFocus();
+  });
+
+  it('closes the popup when clicking the button again', () => {
+    // given
+    const { container, getByRole } = render();
+
+    // when
+    const changeScaleButton = getByRole('button', { name: 'Change Scale' });
+    fireEvent.click(changeScaleButton);
+    fireEvent.click(changeScaleButton);
+
+    // then
+    assertDropdownIsClosed(container);
+  });
+
+  it('closes the popup and disables the button when the connection is lost', () => {
+    // given
+    const { container, getByRole, rerender } = render();
+
+    // when
+    const changeScaleButton = getByRole('button', { name: 'Change Scale' });
+    fireEvent.click(changeScaleButton);
+    rerender({ connected: false });
+
+    // then
+    assertDropdownIsClosed(container);
+    expect(changeScaleButton).toBeDisabled();
+
+    // when
+    rerender({ connected: true });
+
+    // then
+    assertDropdownIsClosed(container);
+    expect(changeScaleButton).toBeEnabled();
+  });
+
+  it('closes the popup when focus is lost', () => {
+    // given
+    const { container, getByRole } = render();
+
+    // when
+    const changeScaleButton = getByRole('button', { name: 'Change Scale' });
+    fireEvent.click(changeScaleButton);
+    fireEvent.blur(getDropdown(container));
+
+    // then
+    assertDropdownIsClosed(container);
+  });
+
+  it('does not close the popup when focus is lost to the button', () => {
+    // given
+    const { container, getByRole } = render();
+
+    // when
+    const changeScaleButton = getByRole('button', { name: 'Change Scale' });
+    fireEvent.click(changeScaleButton);
+    const dropdown = getDropdown(container);
+    fireEvent.blur(dropdown, { relatedTarget: changeScaleButton });
+    fireEvent.blur(changeScaleButton, { relatedTarget: dropdown });
+
+    // then
+    expect(dropdown).toBeVisible();
+  });
+});
+
+const assertDropdownIsClosed = (container: HTMLElement) =>
```

Comment:
Did you consider using `screen.queryByRole` instead of passing the container? This would allow reverting your change to `renderWithWebSocket`.

Reply:
Great idea!